.github/workflows/haskell-ci.yml

```diff
@@ -200,10 +210,14 @@ jobs:
           $CABAL v2-build $ARG_COMPILER $ARG_TESTS $ARG_BENCH --dependencies-only -j2 all
       - name: build
         run: |
-          $CABAL v2-build $ARG_COMPILER $ARG_TESTS $ARG_BENCH all
+          $CABAL v2-build $ARG_COMPILER $ARG_TESTS $ARG_BENCH all --write-ghc-environment-files=always
       - name: tests
         run: |
-          $CABAL v2-test $ARG_COMPILER $ARG_TESTS $ARG_BENCH all
+          $CABAL v2-test $ARG_COMPILER $ARG_TESTS $ARG_BENCH all --test-show-details=direct
+      - name: docspec
+        run: |
+          $CABAL v2-build $ARG_COMPILER $ARG_TESTS $ARG_BENCH all
```

Comment:
Isn't this `$CABAL v2-build $ARG_COMPILER $ARG_TESTS $ARG_BENCH all` invocation redundant? You already run it once above.

Reply:
Unfortunately `v2-test` can mess up the local package db. Not in `lens` case, but I don't know how to know for sure it isn't required.

`cabal` does spend some seconds thinking (it doesn't end up building anything). And whole step takes just 40secs. So I rather not make it more complicated/fragile.

Reply:
I see. But shouldn't `$ARG_TESTS` be `--enable-tests` instead? I fail to see how this could possibly work unless the dependencies from the test suite are included in the build plan.

Reply:
I think you are thinking too much into `lens` example.

The problematic situation I try to avoid here are related to `all` meaning different things for `v2-test` and `v2-build`.

Consider a multi-package project. If some packages don't have `test-suite`s, their libraries won't be built by `v2-test`. For example, in `lens` case `cabal v2-test all --enable-tests` *won't* build `lens-examples`. I think that it's unsafe to assume that `lens-example` will still be registered in a local (project) package db.

However, we build with `v2-build all --enable-tests` just before `v2-test`, so things *seems* to work.

Yet, I don't know (even I should), what `v2-test` and `v2-build` do differently, as `v2-build` after `v2-test` is not a no-op:

```
In order, the following will be built (use -v for more details):
 - lens-5 (lib) (configuration changed)
 - lens-5 (test:doctests) (configuration changed)
 - lens-properties-4.11.1 (lib) (dependency rebuilt)
 - lens-examples-0.1 (exe:lens-pong) (dependency rebuilt)
 - lens-examples-0.1 (lib) (dependency rebuilt)
 - lens-5 (test:templates) (configuration changed)
 - lens-5 (test:properties) (configuration changed)
 - lens-5 (test:hunit) (configuration changed)
Configuring library for lens-5..
Preprocessing library for lens-5..
Building library for lens-5..
Configuring test suite 'doctests' for lens-5..
Preprocessing test suite 'doctests' for lens-5..
Building test suite 'doctests' for lens-5..
Preprocessing library for lens-properties-4.11.1..
Building library for lens-properties-4.11.1..
Preprocessing executable 'lens-pong' for lens-examples-0.1..
Building executable 'lens-pong' for lens-examples-0.1..
Preprocessing library for lens-examples-0.1..
Building library for lens-examples-0.1..
Configuring test suite 'templates' for lens-5..
Warning: The package has an extraneous version range for a dependency on an
internal library: lens >=0 && ==5, lens >=0 && ==5, lens >=0 && ==5, lens >=0
&& ==5, lens >=0 && ==5, lens >=0 && ==5, lens >=0 && ==5, lens >=0 && ==5.
This version range includes the current package but isn't needed as the
current package's library will always be used.
Preprocessing test suite 'templates' for lens-5..
Building test suite 'templates' for lens-5..
Configuring test suite 'properties' for lens-5..
Warning: The package has an extraneous version range for a dependency on an
internal library: lens >=0 && ==5, lens >=0 && ==5, lens >=0 && ==5, lens >=0
&& ==5, lens >=0 && ==5, lens >=0 && ==5, lens >=0 && ==5, lens >=0 && ==5.
This version range includes the current package but isn't needed as the
current package's library will always be used.
Preprocessing test suite 'properties' for lens-5..
Building test suite 'properties' for lens-5..
Configuring test suite 'hunit' for lens-5..
Warning: The package has an extraneous version range for a dependency on an
internal library: lens >=0 && ==5, lens >=0 && ==5, lens >=0 && ==5, lens >=0
&& ==5, lens >=0 && ==5, lens >=0 && ==5, lens >=0 && ==5, lens >=0 && ==5.
This version range includes the current package but isn't needed as the
current package's library will always be used.
Preprocessing test suite 'hunit' for lens-5..
Building test suite 'hunit' for lens-5..
```

It kind of does a lot, but actually does nothing.

So for now, I'd error on the safe side and have an extra `v2-build`.

Reply:
I think you misinterpreted my comment, as I wasn't suggesting to use `v2-test`, but rather using `cabal v2-build --enable-tests` instead of `cabal v2-build --enable-tests`. But in light of https://github.com/ekmett/lens/pull/958/files#r549543219, perhaps the use of `--enable-tests` isn't a good idea.

Reply:
instead of `$ARG_TESTS`?

Unconditional `--enable-tests` may be very bad, e.g. when tests are disabled for genuine reasons, like "a testing lib doesn't work with GHC-x.y" and aren't needed for adding dependencies for `docspec`.